### ISSUE TITLE
Local testing missing packages

### ISSUE THREAD

**Issue by miloush** [dotnet/wpf#3132]
I am trying to test a local WPF build, but since it involves changes in public API, I need the sample app to use the new references rather than replacing published binaries.

I tried to follow the [Developer Guide](https://github.com/dotnet/wpf/blob/master/Documentation/developer-guide.md), basically:

1. forking and cloning the wpf repo
2. importing wpf.vsconfig
3. installing latest .NET Core SDK (5.0.100-preview.5.20279.10 from dotnet.microsoft.com)
4. running `build` from command line, that succeeded
5. running `bulid -test` from administrator command line, that failed, see below
6. changing the code
7. running `build -package`, that succeeded (created 5 files `.nupkg` after build)
8. created a new WPF App (.NET Core)
9. added the `<PropertyGroup>` and `<ItemGroup>`  with `WpfRepoRoot` etc. from _Testing API changes_ section of the _Developer Guide_ at the end of the .csproj file

Still, the Visual Studio does not seem to be using the custom build (does not compile against the API changes). What I noticed is that

```xml
<ReferenceCopyLocalPaths Include="$(WpfRepoRoot)\artifacts\packaging\$(WpfConfig)\Microsoft.DotNet.Wpf.GitHub\lib\$(RuntimeIdentifier)\*.dll" />`
```

does not exist. there is only _netcoreapp5.0_ folder under the _lib_ folders.

Another suspicious thing is the beginning of the `build` output, which looks like this:
```
C:\Users\User\Source\WPF>build -pack
Processing C:\Users\User\Source\WPF\global.json
strawberry-perl was not found in .
net-framework-48-ref-assemblies was not found in .
dotnet-api-docs_netcoreapp3.0 was not found in .
msvcurt-c1xx was not found in .
net-framework-472-iltools was not found in .
Native tools bootstrap failed
Detecting WPF Git hooks...
Detected existing WPF Git pre-commit hook.
Processing C:\Users\User\Source\WPF\eng\..\global.json
strawberry-perl was not found in .
net-framework-48-ref-assemblies was not found in .
dotnet-api-docs_netcoreapp3.0 was not found in .
msvcurt-c1xx was not found in .
net-framework-472-iltools was not found in .
Native tools bootstrap failed
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\MSBuild\Current\Bin\msbuild.exe ...
```

Not sure if that is OK or not.

Possibly related, the `build -test` is failing with
```
...
  PresentationFramework.Aero -> C:\Users\User\Source\WPF\artifacts\bin\PresentationFramework.Aero\Debug\netcoreapp5.0\PresentationFramework.Aero.dll
  The system cannot find the path specified.
C:\Users\User\Source\WPF\eng\WpfArcadeSdk\tools\RunDrtsLocal.targets(10,5): error MSB3073: The command "C:\Users\User\Source\WPF\\artifacts\test\Debug\x86\runte
sts.cmd" exited with code 3. [C:\Users\User\.nuget\packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20309.1\tools\AfterSolutionBuild.proj]

Build FAILED.
```

What did I miss?

**Comment by vatsan-madhavan** [comment]
>strawberry-perl was not found in .
net-framework-48-ref-assemblies was not found in .
dotnet-api-docs_netcoreapp3.0 was not found in .
msvcurt-c1xx was not found in .
net-framework-472-iltools was not found in .
Native tools bootstrap failed
Detecting WPF Git hooks...
Detected existing WPF Git pre-commit hook.
Processing C:\Users\User\Source\WPF\eng\..\global.json
strawberry-perl was not found in .
net-framework-48-ref-assemblies was not found in .
dotnet-api-docs_netcoreapp3.0 was not found in .
msvcurt-c1xx was not found in .
net-framework-472-iltools was not found in .
Native tools bootstrap failed

Those are not failures. They are merely pseudo-failures, and is waiting on https://github.com/dotnet/arcade/issues/4795 (and some subset of [these ](https://github.com/dotnet/arcade/labels/native-toolset-bootstrapping)issues) to be fixed. 😞 

/cc @chcosta, @markwilkie

**Comment by miloush** [comment]
@vatsan-madhavan thought so. I still seem to be blocked though, are there workaround options to test new API? Which process is responsible for generating the `Microsoft.DotNet.Wpf.GitHub\lib\$(RuntimeIdentifier)` bit?

**Comment by vatsan-madhavan** [comment]
That folder is generated by the whole build :-) 

_Testing API Changes_ section is something I’m not apt at troubleshooting. I would defer to the team, or the author of that section.

**Comment by weltkante** [comment]
> Which process is responsible for generating the Microsoft.DotNet.Wpf.GitHub\lib\$(RuntimeIdentifier) bit?

As far as I understood the question the packages are generated but not picked up, right? Maybe just adjust the path to match where the packages actually are generated, docs may be out of date and use the wrong variable.

**Comment by vatsan-madhavan** [comment]
If `build -test` were truly failing, then continuous integration build and PR builds would fail all the time. 

`-pack` should be used always - it's an essential switch. `build -test` might work after a successful `build -pack` - YMMV. Best to call `build -pack -test`. 

If you know that you've only changes one project, subsequent calls to build can take the form `build -pack -test -projects <relative path to .csproj or .vcxproj`>`, IIRC. 

That said, `native tools bootstrap failed` is not a useful part of this or any other problem report. :-/ That leaves us with this: 
```
...
  PresentationFramework.Aero -> C:\Users\User\Source\WPF\artifacts\bin\PresentationFramework.Aero\Debug\netcoreapp5.0\PresentationFramework.Aero.dll
  The system cannot find the path specified.
C:\Users\User\Source\WPF\eng\WpfArcadeSdk\tools\RunDrtsLocal.targets(10,5): error MSB3073: The command "C:\Users\User\Source\WPF\\artifacts\test\Debug\x86\runte
sts.cmd" exited with code 3. [C:\Users\User\.nuget\packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20309.1\tools\AfterSolutionBuild.proj]

Build FAILED.
```

If `PresentationFramework.Aero.dll` is missing, then I worry that you may be attempting to run `-test` before building the repo (fully), and are hoping that `-test` (alone) will be smart enough to do a dependency-based build and _then_ run tests. It won't. Please try `-pack -test` instead. 

In fact `build.cmd` is a wrapper that already incorporates `-restore -build` switches - so what's **really** happening behind the scenes when passing `-pack -test` is that `eng\common\build.ps1` receives `-restore -build -pack -test`.  

If you still get errors, throw in `-verbose` for good measure and try again. 

I'm not sure whether throwing in `-pack` into `build.cmd` (so that developers wouldn't have to bother typing it in all the time) would be benign, but if it were, then it should really be added there. I've always wanted to do this analysis and make that change but never got around to it.

**Comment by miloush** [comment]
@weltkante I am not exactly sure which packages are supposed to be built, but I don't think it's TargetFramework vs RuntimeIdentifier, because the the TargetFramework is imported explicitly earlier, this is the whole section from the documentation:
```xml
  <ItemGroup>
    <Reference Include="$(WpfRepoRoot)\artifacts\packaging\$(WpfConfig)\Microsoft.DotNet.Wpf.GitHub\lib\netcoreapp3.0\*.dll" />
    <ReferenceCopyLocalPaths Include="$(WpfRepoRoot)\artifacts\packaging\$(WpfConfig)\Microsoft.DotNet.Wpf.GitHub\lib\$(RuntimeIdentifier)\*.dll" />
  </ItemGroup>
```

(I already fixed it to netcoreapp5.0)

Another thing I noticed is that while the packaging by means of `-pack` generates few extra `.nuget` files compared to the normal build, the `<ReferenceCopyLocalPaths>` is looking for `*.dll` files, so it might be not related.

@vatsan-madhavan Sorry for confusing you with the PresentationFramework.Aero, that line is actually a success line and there are similar for other libraries above it, the `The system cannot find the path specified.` is the problem.

`-verbose` only increases verbosity of the pre-build process and is not generally helpful, but `-verbosity [minimal]` gets passed to the msbuild.

OK I will dig into the build process, I just thought I must be missing something obvious that everyone else got working.

**Comment by miloush** [comment]
Based on the `RunDrtsLocal.targets`, I believe the continuous integration is actually not running the local DRTs, so this shouldn't have effect on PRs. 

Regarding DRT, this is the offending command:
```
   1>Project "C:\Users\User\.nuget\packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.20309.1\tools\Build.proj" (1) is building "C:\Users\User\.nuget\packages\microsoft.dotnet.arcade.sdk\5.0.0-beta.2030
       9.1\tools\AfterSolutionBuild.proj" (136) on node 2 (Build;Test;Pack target(s)).
136>RunDrts:
         C:\Users\User\Source\WPF\\artifacts\test\Debug\x86\runtests.cmd
         The system cannot find the path specified.
```

The only `runtests.cmd` I have under `artifacts` is this:
```
\WPF\artifacts\packaging\Debug\Microsoft.DotNet.Arcade.Wpf.Sdk.Debug\tools\runtests.cmd
```

I believe this is coming from `\eng\WpfArcadeSdk\tools\RunDrtsLocal.targets`, line 10:
```xml
<Exec Command="$(RepoRoot)\artifacts\test\$(Configuration)\$(Platform)\runtests.cmd"/>
```

The one responsible for creating this seems to be the `CreateTestPayload` target in  `\eng\WpfArcadeSdk\tools\CreateTestPayload.targets` but I don't see it called in the build log.

That target should be called by `\eng\WpfArcadeSdk\Sdk\Sdk.targets`, however, it is conditioned!
```xml
  <Import Project="$(WpfArcadeSdkToolsDir)CreateTestPayload.targets" Condition="'$(CreateTestPayload)'=='true'"/>
```

Trying `build /p:CreateTestPayload=true` fails all the projects with 

```
\eng\WpfArcadeSdk\tools\CreateTestPayload.targets(48,5): error MSB3030: Could not copy the file "BranchCommon\data\DrtList.xml" because it was not found.
```

These errors are coming from line 48 of `CreateTestPayload.targets`:
```
<Copy SourceFiles="$(WpfTestsDir)BranchCommon\data\DrtList.xml"
          DestinationFolder="$(WpfTestBasePayloadPath)DRT"  />
```

The `WpfTestsDir` is set by `\src\Microsoft.DotNet.Wpf\test\Directory.Build.props` so it seems the `CreateTestPayload` target has dependency on this project. I also found this target mentioned in (Testing in Helix)[https://github.com/dotnet/wpf/blob/master/Documentation/testing-in-helix.md] which however not that helpful in resolving the issue, at least to me. It does say that `DrtXaml` is the one responsible for creating the infrastructure using `CreateTestPayload`.

Nevertheless, the `runtests.cmd` file we were missing at the beginning gets created this way. However, it fails with `System.IO.DirectoryNotFoundException: \artifacts\test\Debug\x86\Test\DRT could not be found. Did you forget to build it?` These should get copied over by `AddDrtsToPayload` but that only includes `DrtXaml` as far as I can tell. Hmmm... isn't that like a dependency cycle?

I guess the documentation needs to be updated with how to setup the tests for the first time. I will put this line of experiments on hold and let someone experienced with the test infrastructure to chime in, as that is not really blocking me.

**Comment by miloush** [comment]
As for `native tools bootstrap failed`, that is just a an artefact of `\eng\common\native\install-tool.ps1` non-sensical check for a unique `$ToolName.exe` in the installed tool subtree. For example, it's looking for `net-framework-472-iltools.exe` inside `\.tools\native\bin\net-framework-472-iltools\0.0.0.1\`, but that's just not how any of this works.

I will assume the shims are actually not needed.

**Comment by vatsan-madhavan** [comment]
> not exactly sure which packages are supposed to be built

- Microsoft.DotNet.Arcade.Wpf.Sdk
- Microsoft.DotNet.Wpf.GitHub
- Microsoft.DotNet.Wpf.ProjectTemplates
- Microsoft.NET.Sdk.WindowsDesktop

They may each have a `.Debug` suffix sometimes. 

Building with `-pack` runs `PreparePackageAssets` target during the build, which is important for a lot of things. Importantly, it lays down (and updates) binaries and PDB's under `artifacts\packaging\` in a normalized structure, which is useful in a variety of ways and easier to use than relying on the `artifacts\obj` hierarchy.

**Comment by vatsan-madhavan** [comment]
>I will assume the shims are actually not needed.

they aren't needed. 
https://github.com/dotnet/arcade/pull/2249

**Comment by vatsan-madhavan** [comment]
There doesn't seem to be any _trick_ to running the DRT's - they are just not launching. They are not launching because the tests aren't included in the solution. 

You need this change to get the tests back in the main solution - https://github.com/wpfcontrib/wpf/commit/f458cd599f1cf199d94aa6eec4f6510c329a666a. 

That gets the tests running, but they crashed during execution for me. YMMV.
